staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/filter.go

```diff
@@ -156,9 +160,22 @@ func (f *filter) ForInput(versionedAttr *generic.VersionedAttributes, versionedP
 			continue
 		}
 		t1 := time.Now()
-		evalResult, _, err := compilationResult.Program.Eval(va)
+		evalResult, evalDetails, err := compilationResult.Program.Eval(va)
 		elapsed := time.Since(t1)
 		evaluation.Elapsed = elapsed
+		if evalDetails == nil {
+			return nil, errors.New(fmt.Sprintf("runtime cost could not be calculated for expression: %v, no further expression will be run", compilationResult.ExpressionAccessor.GetExpression()))
+		} else {
+			rtCost := evalDetails.ActualCost()
+			if rtCost == nil {
+				return nil, errors.New(fmt.Sprintf("runtime cost could not be calculated for expression: %v, no further expression will be run", compilationResult.ExpressionAccessor.GetExpression()))
+			} else {
+				if *rtCost > math.MaxInt64 || int64(*rtCost) > remainingBudget {
+					return nil, errors.New(fmt.Sprintf("validation failed due to running out of cost budget, no further validation rules will be run"))
+				}
+				remainingBudget -= int64(*rtCost)
+			}
+		}
```

Comment:
Note: here we make the behavior same as the CRD validation rule path, which is whenever issues in retrieving cost, per expression cost exceed or runtime budge exceed, we stop running following expressions and return error. 
However, it will make the cost exceed error higher priority than other errors and return only per expression exceed error back even there are another validation failures existing.
Should we treat the per expression cost exceed error same as other errors and save per `evaluationResult`? And we could also consider keeping the following validations running. What do you think? @jpbetz

Reply:
I think it's fine to only return one per expression exceeded error back. Users shouldn't expect to get full information back when the cost system kicks in, since the point of the cost system is to prevent running CEL expressions once the limit is exceeded.